<a href="https://colab.research.google.com/github/maynak-dev/Machine-Learning/blob/main/Imbalanced_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Imbalanced DataSet:
A dataset with an unequal class distribution

In [1]:
# Importing the dependencies
import numpy as np
import pandas as pd

In [2]:
# Loading the dataset to pandas dataframe
credit_card_data = pd.read_csv('/content/credit_data.csv')

In [3]:
# printing first 5 rows of the dataframe
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
# Printing last 5 rows of the dataframe
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [5]:
# distribution of the two classes
credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

0 --> Legit Transaction

1 --> Fraudulent Transaction

Here almost 99% Legit transaction as the number of legit transaction is 284315 and fraud is only 492 therefore imbalanced dataset cause we have more number of datapoints for a particular class.

In [6]:
# separating the legit and fraudulent transaction
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [7]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


**Implementing Under-Sampling Technique**

Build a sample dataset containing similar distribution of Legit and Fraudulent transaction.

No. od Fraudulent transaction --> 492

In [8]:
#undersampling technique
legit_sample = legit.sample(n=492)

In [9]:
print(legit_sample.shape)

(492, 31)


In [10]:
# Concatenate the two dataframes
new_dataset = pd.concat([legit_sample, fraud], axis=0)  # axis = 0 means we want to concatenate row wise

In [11]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
281879  170486.0  2.148201 -0.770551 -0.781040 -0.315908 -0.963826 -0.973288   
165646  117560.0  1.909263  0.180221 -0.675716  3.618746  0.683523  1.495493   
216159  140364.0 -2.908769 -2.224978  1.538665 -1.741638  0.668993 -1.596584   
98095    66516.0  1.161615  0.164145  0.390606  1.398020 -0.193216 -0.188899   
260721  159693.0  1.960068 -0.152349 -1.668237  0.165012  0.445980 -0.076384   

              V7        V8        V9  ...       V21       V22       V23  \
281879 -0.593445 -0.252055 -0.177017  ... -0.254847 -0.084298  0.215065   
165646 -0.416516  0.382756 -0.344693  ...  0.068401  0.387454  0.009698   
216159 -1.506531  0.812955 -1.106280  ...  0.325630 -0.222530 -0.157405   
98095   0.034291  0.014894  0.371240  ... -0.107275 -0.101639 -0.057729   
260721 -0.160305 -0.007831  0.449622  ... -0.084646 -0.208920  0.121000   

             V24       V25       V26       V27       V28  Amount  Class  
281879  0.062700 -0.241694  0.688255 -0.037387 -0.056063    8.60      0  
165646 -0.223143  0.169961  0.143453 -0.014577 -0.061246    1.51      0  
216159 -0.013839  0.356836 -0.421105  0.170027 -0.334058   39.64      0  
98095   0.074899  0.645447 -0.310866  0.036100  0.015133   10.03      0  
260721  0.039975 -0.193939  0.494676 -0.059635 -0.031222   48.45      0  

[5 rows x 31 columns]

In [12]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

As we can see in the Class column that the legit transaction goes upward and the fraudulent transactions added to the bottom.

In [13]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64